### DISCUSSION TITLE
Shrink and smooth palletized image, but preserve outer edges?

### DISCUSSION THREAD

**Discussion by pw32x** [ImageMagick/ImageMagick#5157] [Help]
Does ImageMagick have a mode to shrink and smooth a low resolution palletized image which also preserves the edges so that it doesn't blend with the designated background color?

Background:

I have a 128x128 256-color image of a sprite and I'm building a chain of smaller images from it (64x64, 32x32, 16x16). In my project, the black background color (index 0) is interpreted as being transparent. Images are meant to be smoothed out when shrunk, but unfortunately it also smooths the edges which leaves a dark outline when displayed on screen. 

(small image warning)

Original 128x128 image. 
![lod0](https://user-images.githubusercontent.com/96802526/168713276-65c8ace1-2099-4b6f-a22c-517baf1b9ae9.png)

64x64 shrunk and smoothed image. The image preserves the 256 color palette. No new colors are created.
![lod1](https://user-images.githubusercontent.com/96802526/168713291-214bfb5e-2002-4e39-b029-6bc2e3ea3909.png)

Same 64x64 but black background made transparent as seen in my project. Notice the dark outline.
![lod1_edge](https://user-images.githubusercontent.com/96802526/168713299-56e3196a-71d3-4ebd-990a-4c907d3d10c8.png)

You can see the dark outline in motion in this gif (middle sprite):
https://twitter.com/pw_32x/status/1526020681728663558?s=20&t=z2vwr4UTyEnmuQ2Wo5Q-wQ
(Left image is the original 128x128, middle image 128x128 with LODs/mips, right image 128x128 with LODs/mips and manual fixing of dark edges.)

Thanks!

**Comment by fmw42** [comment]
Try -sample .... It will shrink by skipping pixels. That way there is not blending of neighboring pixels colors. 

Or try -filter point -resize ...

See

https://imagemagick.org/script/command-line-options.php#sample

**Comment by snibgo** [comment]
>  In my project, the black background color (index 0) is interpreted as being transparent.

Most background pixels are very dark blue, but not black.

Between the very dark blue background and the green trees, you have pixels that are between those colours, ie fairly dark blue. These are probably from anti-aliasing. You can make the background transparent, but the fairly dark blue pixels give "a dark outline when displayed on screen."

See the recent thread https://github.com/ImageMagick/ImageMagick/discussions/5146 , which has the same problem.

**Comment by pw32x** [comment]
Thanks for the answers guys, but I think my question was over complicated.

The question I should've originally asked is: how can I smooth the inside of an image without smoothing the edges?

In the image below I've used a red background to better show the effect I'm seeing (picture 4) and the result I'm looking for (picture 5)
![example](https://user-images.githubusercontent.com/96802526/168815036-c6b5de04-e433-446e-9128-26b549c54ab6.png)

Picture 5 was created manually by combining the inside of the smoothed version with the edges of the unsmoothed version.

**Reply by fmw42** [reply]
Make a mask of the inside vs outside. Resize with smoothing (e.g. -scale or -filter X -resize) on the whole image. Then use the mask (and use morphology erode to shrink the mask one pixel around the border) to combine the original with the smoothed image  so that the inside is smoothed and the outside is not.

**Comment by snibgo** [comment]
@pw32x: What commands are you using to "shrink without smoothing" and "shrink with smoothing"?

**Reply by pw32x** [reply]
I'm not using any commands, actually. The smoothing in the image above is done manually (through Graphics Gale) to illustrate what I want to do. I know IM can both smooth and shrink images so I'm not worried about that part. It's the preservation of the edges that I don't know how to tackle.

**Comment by fmw42** [comment]
( Please always provide your IM version and platform when asking questions)

Does this do what you want?  Unix syntax for Imagemagick 6.  This is what I suggested above. I shrink the input using -sample to avoid blurring. I then shrink using -scale to blur by averaging 2x2 blocks. I then create a mask from the first image and erode it slightly. Then I composite the two reduced images using the mask.

![168713276-65c8ace1-2099-4b6f-a22c-517baf1b9ae9](https://user-images.githubusercontent.com/13458218/168864730-ee10d3ee-b734-422c-990c-5501a88fcca2.png)


```
convert 168713276-65c8ace1-2099-4b6f-a22c-517baf1b9ae9.png -write mpr:img +delete \
\( mpr:img -sample 50x50% \) \
\( mpr:img -scale 50x50% \) \
\( -clone 0 -threshold 3.2% +write x1.png -morphology erode diamond:1 +write x2.png \) \
-compose over -composite \
168713276-65c8ace1-2099-4b6f-a22c-517baf1b9ae9_reduced.png

```

![168713276-65c8ace1-2099-4b6f-a22c-517baf1b9ae9_reduced](https://user-images.githubusercontent.com/13458218/168864764-b357a8a0-8486-419f-a7fa-a3d055656bd1.png)

**Reply by pw32x** [reply]
That sounds feasible! I'll take a look, thanks!
